drivers/gpio/gpio_intel_socfpga.c

```diff
@@ -0,0 +1,305 @@
+#define GPIO_INTEL_RESET_SUPPORT	(DT_INST_FOREACH_STATUS_OKAY(GPIO_INTEL_CHECK_RESET) 0)
+
+#define DEV_CFG(_dev) ((const struct gpio_intel_socfpga_config *)(_dev)->config)
+#define DEV_DATA(_dev) ((struct gpio_intel_socfpga_data *const)(_dev)->data)
+
+#define GPIO_SWPORTA_DDR_OFFSET	0x04
+#define GPIO_EXT_PORTA_OFFSET	0x50
+#define GPIO_ID_CODE_OFFSET	0x64
+#define GPIO_VER_ID_CODE_OFFSET	0x6c
+
+#define PMUX_SEL_OFFSET		4
+#define PMUX_40_SEL		40
+#define PMUX_40_SEL_GAP		96
+#define PMUX_GPIO_VAL		8
+
+struct gpio_intel_socfpga_config {
+	/** GPIO common config */
+	struct gpio_driver_config gpio_config;
+	/** GPIO MMIO mapped physical address */
+	DEVICE_MMIO_NAMED_ROM(gpio_mmio);
+	/** GPIO port number (0 or 1) */
+	const uint32_t gpio_port;
+	/** GPIO pin number per port */
+	uint32_t ngpios;
+#if GPIO_INTEL_RESET_SUPPORT
+	/** GPIO reset device information */
+	struct reset_dt_spec reset;
+#endif
+};
+
+struct gpio_intel_socfpga_data {
+	/** GPIO common config */
+	struct gpio_driver_data gpio_data;
+	/** GPIO MMIO mapped virtual address */
+	DEVICE_MMIO_NAMED_RAM(gpio_mmio);
+};
+
+/**
+ * @brief provides funtionality to validate pin/pinmask before gpio operation
+ *
+ * @param cfg GPIO port instance configuration.
+ * @param pin_mask GPIO pin number/numbers.
+ *
+ * @retval true on success.
+ * @retval false if pin/pinmask is not valid.
+ */
+static bool gpio_socfpga_is_pinmask_valid(
+		const struct gpio_intel_socfpga_config *cfg,
+		uint32_t pin_mask)
+{
+	const gpio_port_pins_t port_pin_mask = cfg->gpio_config.port_pin_mask;
+	const uint32_t gpio_port = cfg->gpio_port;
+	uint32_t ngpios = cfg->ngpios;
+
+	uint32_t pin;
+	uint32_t hps_io_pin;
+	uint32_t addroffst;
+	uint32_t pmux_value;
+
+	/* Pin range more than 32 will return 0x0 pin mask */
+	if (pin_mask == 0) {
+		LOG_DBG("Invalid pin_mask value passed: 0x%X", pin_mask);
+		return false;
+	}
+	/**
+	 * This loop will verify every bit set as a separate pin.
+	 * Each pin should be eligible for gpio operation in pin_mask
+	 */
+	while (pin_mask != 0) {
+		pin = find_lsb_set(pin_mask) - 1;
+
+		/* Check whether input pin is in range */
+		if ((BIT(pin) & port_pin_mask) == 0) {
+			LOG_DBG("Mask bit %d value is out of range: MASK:0x%X", pin, port_pin_mask);
+			return false;
+		}
+
+		/* Check for GPIO pinmux configuration value */
+		hps_io_pin = (gpio_port * ngpios) + pin;
+		addroffst = hps_io_pin * PMUX_SEL_OFFSET;
+
+		/**
+		 * The pinmux register has address jump in hardware
+		 * Address is not incremented from 0x9C to 0xA0. See below details.
+		 */
+		/**
+		 * pin 0 -  0x10D1 3000
+		 * pin 1 -  0x10D1 3004
+		 * ...
+		 * pin 39 - 0x10D1 309C
+		 * (Jump)
+		 * pin 40 - 0x10D1 3100
+		 * ...
+		 * Below GAP is added after pin 40.
+		 */
+		if (hps_io_pin >= PMUX_40_SEL) {
+			addroffst += PMUX_40_SEL_GAP;
+		}
+
+		/* Getting pinmux base address from pinmux node */
+		pmux_value = sys_read32((DT_REG_ADDR(DT_NODELABEL(pinmux)) + addroffst));
```

Comment:
I'd avoid using a these macros deep in the code as it hides the dependency.

Reply:
In understand your main concern related to complete pinmux/pinctrl implementation. 
Currently another hardware outside zephyr world, configuring pin controller register on our platform. 
Our main design goal is to make sure pinctrl is configured in GPIO mode before performing any GPIO operation. 
Also, Unfortunately, Pinmux driver does not have get pinctrl register values. If you say, I can create a RFC for pinctrl enhancement separately. Is that fine ?